relations/src/shielder/deposit.rs

```diff
@@ -1,149 +1,96 @@
 use ark_ff::BigInteger256;
 use ark_r1cs_std::alloc::AllocVar;
-use ark_relations::{
-    ns,
-    r1cs::{
-        ConstraintSynthesizer, ConstraintSystemRef, SynthesisError,
-        SynthesisError::AssignmentMissing,
-    },
-};
-use ark_std::{marker::PhantomData, vec, vec::Vec};
-
-use super::{
-    note::check_note,
-    types::{
-        BackendNote, BackendNullifier, BackendTokenAmount, BackendTokenId, BackendTrapdoor,
-        FrontendNote, FrontendNullifier, FrontendTokenAmount, FrontendTokenId, FrontendTrapdoor,
-    },
-};
-use crate::{
-    environment::{CircuitField, FpVar},
-    relation::{
-        state::{FullInput, NoInput, OnlyPublicInput, State, WithPublicInput},
-        GetPublicInput,
-    },
-};
+use ark_relations::ns;
+use snark_relation_proc_macro::snark_relation;
+
+use crate::{BackendNote, FrontendNote};
+
+fn convert_note(front: FrontendNote) -> BackendNote {
+    BackendNote::from(BigInteger256::new(front))
+}
```

Comment:
Could this be turned into an `Into`/`From` implementation?

And, by extend, would it work if we had `mod backend { ... }` and `mod frontend { ... }` with representations and mappings (as `From`/`Into` impls) between the two? Then, maybe we could cut down on the "boilerplate" in the `#[public_input(...)]` macro and just require that there exists `BackendType: Into<FrontedType>` instead of `#[public_input(frontend_type = "FrontendNote", parse_with = "convert_note")]`

Reply:
into/from won't work, since we cannot implement foreign traits for foreign types (notice that all front/backend types are just aliases)

I was trying to make a step towards it here: https://github.com/Cardinal-Cryptography/HouseSnark/pull/35/files, but back then there was _a lot_ of code to implement strong typing in full... however, I would revisit this somewhere soon

Reply:
We could add our own mapping trait then and implement that instead (of `Into`/`From`).

Reply:
yeap, that's the easiest alternative, however, slightly out of scope here

Reply:
Yes, just mentioning the possible improvement to the macro.

Reply:
talking about improvements, it should be fairly easy to add generics support as well